capsules/src/log_storage.rs

```diff
@@ -0,0 +1,865 @@
+    ///     * ERESERVE: client or internal pagebuffer missing.
+    fn read_entry_header(&self, entry_cookie: usize) -> Result<usize, ReturnCode> {
+        self.pagebuffer
+            .take()
+            .map_or(Err(ReturnCode::ERESERVE), move |pagebuffer| {
+                // Get length.
+                const LENGTH_SIZE: usize = size_of::<usize>();
+                let length_bytes = self.get_bytes(entry_cookie, LENGTH_SIZE, pagebuffer);
+                let length_bytes = <[u8; LENGTH_SIZE]>::try_from(length_bytes).unwrap();
+                let length = usize::from_ne_bytes(length_bytes);
+
+                // Return length of next entry.
+                self.pagebuffer.replace(pagebuffer);
+                if length == 0 || length > self.page_size - PAGE_HEADER_SIZE - ENTRY_HEADER_SIZE {
+                    Err(ReturnCode::FAIL)
+                } else {
+                    Ok(length)
+                }
+            })
+    }
+
+    /// Reads the next entry into a buffer. Returns the number of bytes read on success, or an
+    /// error otherwise.
+    /// ReturnCodes used:
+    ///     * FAIL: reached end of log, nothing to read.
+    ///     * ERESERVE: internal pagebuffer missing, log is presumably broken.
+    ///     * ESIZE: buffer not large enough to contain entry being read.
+    fn read_entry(&self, buffer: &mut [u8], length: usize) -> Result<usize, ReturnCode> {
+        // Get next entry to read. Immediately returns FAIL in event of failure.
+        let entry_cookie = self.get_next_entry()?;
+        let entry_length = self.read_entry_header(entry_cookie)?;
+
+        // Read entry into buffer.
+        self.pagebuffer
+            .take()
+            .map_or(Err(ReturnCode::ERESERVE), move |pagebuffer| {
+                // Ensure buffer is large enough to hold log entry.
+                if entry_length > length {
+                    self.pagebuffer.replace(pagebuffer);
+                    return Err(ReturnCode::ESIZE);
+                }
+                let entry_cookie = entry_cookie + ENTRY_HEADER_SIZE;
+
+                // Copy data into client buffer.
+                let data = self.get_bytes(entry_cookie, entry_length, pagebuffer);
+                for i in 0..entry_length {
+                    buffer[i] = data[i];
+                }
+
+                // Update read cookie and return number of bytes read.
+                self.read_cookie.set(entry_cookie + entry_length);
+                self.pagebuffer.replace(pagebuffer);
+                Ok(entry_length)
+            })
+    }
+
+    /// Writes an entry header at the given cookie within a page. Must write at most
+    /// ENTRY_HEADER_SIZE bytes.
+    fn write_entry_header(&self, length: usize, cookie: usize, pagebuffer: &mut F::Page) {
+        let mut offset = 0;
+        for byte in &length.to_ne_bytes() {
+            pagebuffer.as_mut()[cookie + offset] = *byte;
+            offset += 1;
+        }
+
+        assert!(offset <= ENTRY_HEADER_SIZE);
```

Comment:
given that ENTRY_HEADER_SIZE is defined as sizeof(usize), and this for loop iterates through the number of bytes in a usize, this assert statement seems superfluous